include/nbl/builtin/hlsl/bxdf/ndf/ggx.hlsl

```diff
@@ -256,19 +235,224 @@ struct GGX<T,true NBL_PARTIAL_REQ_BOT(concepts::FloatingPointScalar<T>) >
+    enable_if_t<C::value && IsBSDF, quant_type> D(NBL_CONST_REF_ARG(quant_query_type) quant_query, NBL_CONST_REF_ARG(LS) _sample, NBL_CONST_REF_ARG(Interaction) interaction, NBL_CONST_REF_ARG(MicrofacetCache) cache)
+    {
+        scalar_type d = __ndf_base.template D<MicrofacetCache>(cache);
+        quant_type dmq;
+        dmq.microfacetMeasure = d;  // note: microfacetMeasure/2NdotV
+
+        const scalar_type VdotHLdotH = quant_query.getVdotHLdotH();
+        const scalar_type VdotH_etaLdotH = quant_query.getVdotH_etaLdotH();
+        const bool transmitted = reflect_refract==MTT_REFRACT || (reflect_refract!=MTT_REFLECT && VdotHLdotH < scalar_type(0.0));
+        scalar_type NdotL_over_denominator = _sample.getNdotL(BxDFClampMode::BCM_ABS);
+        if (transmitted)
+                NdotL_over_denominator *= -scalar_type(4.0) * VdotHLdotH / (VdotH_etaLdotH * VdotH_etaLdotH);
+        dmq.projectedLightMeasure = d * NdotL_over_denominator;
+        return dmq;
+    }
+
+    template<class LS, class Interaction, typename C=bool_constant<!IsBSDF> NBL_FUNC_REQUIRES(LightSample<LS> && RequiredInteraction<Interaction>)
+    enable_if_t<C::value && !IsBSDF, quant_type> DG1(NBL_CONST_REF_ARG(dg1_query_type) query, NBL_CONST_REF_ARG(quant_query_type) quant_query, NBL_CONST_REF_ARG(LS) _sample, NBL_CONST_REF_ARG(Interaction) interaction)
+    {
+        scalar_type dg1 = base_type::template DG1<dg1_query_type>(query);
+        quant_type dmq;
+        dmq.microfacetMeasure = dg1;
+        dmq.projectedLightMeasure = dg1;// TODO: figure this out * _sample.getNdotL(BxDFClampMode::BCM_MAX);
+        return dmq;
+    }
+    template<class LS, class Interaction, typename C=bool_constant<IsBSDF> NBL_FUNC_REQUIRES(LightSample<LS> && RequiredInteraction<Interaction>)
+    enable_if_t<C::value && IsBSDF, quant_type> DG1(NBL_CONST_REF_ARG(dg1_query_type) query, NBL_CONST_REF_ARG(quant_query_type) quant_query, NBL_CONST_REF_ARG(LS) _sample, NBL_CONST_REF_ARG(Interaction) interaction)
+    {
+        scalar_type dg1 = base_type::template DG1<dg1_query_type>(query);
+        quant_type dmq;
+        dmq.microfacetMeasure = dg1;  // note: microfacetMeasure/2NdotV
+
+        const scalar_type VdotHLdotH = quant_query.getVdotHLdotH();
+        const scalar_type VdotH_etaLdotH = quant_query.getVdotH_etaLdotH();
+        const bool transmitted = reflect_refract==MTT_REFRACT || (reflect_refract!=MTT_REFLECT && VdotHLdotH < scalar_type(0.0));
+        scalar_type NdotL_over_denominator = _sample.getNdotL(BxDFClampMode::BCM_ABS);
+        if (transmitted)
+                NdotL_over_denominator *= -scalar_type(4.0) * VdotHLdotH / (VdotH_etaLdotH * VdotH_etaLdotH);
+        dmq.projectedLightMeasure = dg1;// TODO: figure this out * NdotL_over_denominator;
+        return dmq;
+    }
```

Comment:
This stuff maybe we can rethink, I'd be tempted to allow an NDF "skip" providing `D` and `correlated` if it has both the `DG1` and `Dcorrelated` methods (so the cook torrance can leverage that and skip multiplying `D` with `correlated` by itself for `eval`)